### PR TITLE
Add MarkHistoryItemAsFailed function to Sonarr

### PR DESCRIPTION
**Pull request by JackDallas**
Very open to changing the function name and details, I just needed this for my project so I added it :+1: 

Let me know any comments you have

### CODE REVIEW

**Comment by austinwbest** [comment]
I would suggest putting this across all applicable arr apps instead of just one :)

File: go.mod [hunk 1]
```diff
@@ -1,4 +1,4 @@
-module golift.io/starr
+module github.com/jackdallas/starr
```
Comment (davidnewhall) [inline go.mod:1]: You should use `replace =>` in your `go.mod` instead of doing this. Can't merge this as-is.

File: sonarr/sonarr.go [hunk 2]
```diff
@@ -398,3 +398,19 @@ func (s *Sonarr) GetHistory(maxRecords int) (*History, error) {
 
 	return &history, nil
 }
+
+// MarkHistoryItemAsFailed marks the given history item as failed by id.
+func (s *Sonarr) MarkHistoryItemAsFailed(id int64) error {
```
Comment (davidnewhall) [inline sonarr/sonarr.go:403]: Let's rename the method to `Fail()`. simple and ez.

What actually this ID? Is it a specific `historyID`? Is it a `seriesID`? Perhaps it's an `episodeID`? Can you rename `id` to be more descriptive?

**Review by davidnewhall — CHANGES_REQUESTED** [review]
@austinwbest This endpoint is not documented. I don't see it documented for any of the starr apps. That's a problem when asking someone to implement it on apps they don't use or aren't familiar with. I will accept this for just Sonarr since that's the case. 
@JackDallas - Please see my other comments.

File: sonarr/sonarr.go (hunk 2 above)
Comment (JackDallas) [reply sonarr/sonarr.go]: Renamed and I updated the `id` to be `historyID` 👍

File: go.mod (hunk 1 above)
Comment (JackDallas) [reply go.mod]: Hey my bad I meant to push this to another branch temporarily, cleaned up :+1:

**Comment by JackDallas** [comment]
Sorry for the delay @davidnewhall I've made the changes asked for and cleaned up the commits (Meant for them to go to a temp branch, clearly hadn't had enough coffee that day)

File: radarr/radarr.go [hunk 3]
```diff
@@ -431,3 +431,19 @@ func (r *Radarr) SendCommand(cmd *CommandRequest) (*CommandResponse, error) {
...
+
+// MarkHistoryItemAsFailed marks the given history item as failed by id.
+func (r *Radarr) Fail(historyID int64) error {
+	if historyID < 1 {
+		return fmt.Errorf("invalid history ID: %d", historyID)
+	}
+
+	params := make(url.Values)
+
+	_, err := r.Post("v3/history/failed/"+strconv.FormatInt(historyID, starr.Base10), params, nil)
```
Comment (davidnewhall) [inline radarr/radarr.go:443]: Is this supposed to be a Post? The body is `nil`, and that seems suspect.
If it's supposed to a Post please fix the `Errorf` line below that says `api.Get` (it's a Post not a Get).

Same for Sonarr changes.

File: radarr/radarr.go [hunk 4]
```diff
@@ -431,3 +431,19 @@ func (r *Radarr) SendCommand(cmd *CommandRequest) (*CommandResponse, error) {
 
 	return &output, nil
 }
+
+// MarkHistoryItemAsFailed marks the given history item as failed by id.
+func (r *Radarr) Fail(historyID int64) error {
+	if historyID < 1 {
+		return fmt.Errorf("invalid history ID: %d", historyID)
```
Comment (davidnewhall) [inline radarr/radarr.go:438]: Let's avoid dynamic errors. This triggered a [build error.](https://app.travis-ci.com/github/golift/starr/builds/242611275) 
Recommend:
```golang
fmt.Errorf("%w: invalid history ID: %d", ErrRequestError, historyID)
```

https://github.com/golift/starr/blob/a5e9b3f34ec7015a769e4b95fd4c5254a13230c2/radarr/radarr.go#L254-L255

However, see my next comment about this problem in Sonarr, it requires a slight change here too.

File: sonarr/sonarr.go [hunk 5]
```diff
@@ -398,3 +398,19 @@ func (s *Sonarr) GetHistory(maxRecords int) (*History, error) {
 
 	return &history, nil
 }
+
+// Fail marks the given history item as failed by id.
+func (s *Sonarr) Fail(historyID int64) error {
+	if historyID < 1 {
+		return fmt.Errorf("invalid history ID: %d", historyID)
```
Comment (davidnewhall) [inline sonarr/sonarr.go:405]: The `ErrRequestError` variable does not exist in the Sonarr package, so perhaps lets move that variable into `starr/shared.go` and change the ref in Radarr, and add a ref here.
```golang
fmt.Errorf("%w: invalid history ID: %d", starr.ErrRequestError, historyID)
```

**Review by davidnewhall — CHANGES_REQUESTED** [review]
Thanks for the updates. Added a few more requests.

**Comment by JackDallas** [comment]
Hey @davidnewhall I will be addressing your comments, but I've realised since that getting history has an issue as the max results you can request seems to be limited to 250, I'm going to implement pagination to abstract that away when I have a chance and upload that with the requested changes :+1:

**Comment by davidnewhall** [comment]
Sounds good to me. I appreciate you putting some time into this!

**Comment by davidnewhall** [comment]
Sometimes you can just pass a `pageSize` param to up the limit? I do that instead of paginate. Later I'd like to add a pagination function to wrap the normal function, much like how the AWS library works.

https://github.com/golift/starr/blob/master/radarr/radarr.go#L81

**Comment by davidnewhall** [comment]
I never ran into a limit of 250. I was able to pull all of history items in a single request, or in a series of pages. I rewrote the methods for getting history and queue. Added pagination support. It all works behind the scenes.  That's in #20.

I then merged this, and I'm updating the methods to be consistent. See #21 for that.

**Comment by JackDallas** [comment]
@davidnewhall my apologies life caught up with me these past few months, I hope the changes I provided could of been of some help, thanks for the awesome project!

**Comment by davidnewhall** [comment]
No worries Jack. :) Your changes are appreciated!

**Pull request merged**